syntax = "proto3";

package cinema;

option java_package = "com.movieticket.cinema.grpc";
option java_outer_classname = "CinemaServiceProto";

// Cinema gRPC Service Definition
service CinemaService {
  // Check seat availability for a specific showtime
  rpc CheckSeatAvailability(SeatAvailabilityRequest) returns (SeatAvailabilityResponse);
  
  // Lock seats for booking (with timeout)
  rpc LockSeats(LockSeatsRequest) returns (LockSeatsResponse);
  
  // Release locked seats (in case of booking failure)
  rpc ReleaseSeatLock(ReleaseSeatLockRequest) returns (ReleaseSeatLockResponse);
  
  // Confirm seat booking (finalize the reservation)
  rpc ConfirmSeatBooking(ConfirmSeatBookingRequest) returns (ConfirmSeatBookingResponse);
  
  // Get showtime details
  rpc GetShowtimeDetails(ShowtimeDetailsRequest) returns (ShowtimeDetailsResponse);
}

// Request/Response Messages

message SeatAvailabilityRequest {
  string showtime_id = 1;
  repeated string seat_numbers = 2;
}

message SeatAvailabilityResponse {
  bool available = 1;
  repeated SeatInfo unavailable_seats = 2;
  string message = 3;
}

message LockSeatsRequest {
  string showtime_id = 1;
  repeated string seat_numbers = 2;
  string booking_id = 3;
  int32 lock_duration_seconds = 4; // Default: 300 seconds (5 minutes)
}

message LockSeatsResponse {
  bool success = 1;
  string lock_id = 2;
  int64 expires_at = 3; // Unix timestamp
  repeated string failed_seats = 4;
  string message = 5;
}

message ReleaseSeatLockRequest {
  string lock_id = 1;
  string booking_id = 2;
}

message ReleaseSeatLockResponse {
  bool success = 1;
  string message = 2;
}

message ConfirmSeatBookingRequest {
  string lock_id = 1;
  string booking_id = 2;
  string user_id = 3;
}

message ConfirmSeatBookingResponse {
  bool success = 1;
  repeated string confirmed_seats = 2;
  string message = 3;
}

message ShowtimeDetailsRequest {
  string showtime_id = 1;
}

message ShowtimeDetailsResponse {
  ShowtimeInfo showtime = 1;
  MovieInfo movie = 2;
  CinemaInfo cinema = 3;
}

// Data Transfer Objects

message SeatInfo {
  string seat_number = 1;
  SeatStatus status = 2;
  string locked_by = 3; // booking_id if locked
  int64 locked_until = 4; // Unix timestamp
}

message ShowtimeInfo {
  string id = 1;
  string movie_id = 2;
  string cinema_id = 3;
  int64 start_time = 4; // Unix timestamp
  int64 end_time = 5; // Unix timestamp
  double base_price = 6;
  int32 total_seats = 7;
  int32 available_seats = 8;
}

message MovieInfo {
  string id = 1;
  string title = 2;
  string genre = 3;
  int32 duration_minutes = 4;
  string rating = 5;
  string poster_url = 6;
}

message CinemaInfo {
  string id = 1;
  string name = 2;
  string location = 3;
  int32 total_screens = 4;
}

// Enums

enum SeatStatus {
  AVAILABLE = 0;
  LOCKED = 1;
  BOOKED = 2;
  MAINTENANCE = 3;
}